torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp

```diff
@@ -17,171 +18,881 @@ namespace fuser {
+  // also copied from SchedulerRuntimeInfo::getMaxVectorizableWidth
+  bool use_root_dom = of_tv->hasReduction() && of_tv->hasRFactor();
+  auto of_tv_root =
+      use_root_dom ? of_tv->getRootDomain() : of_tv->getMaybeRFactorDomain();
+
+  const auto& projected_dim_map =
+      use_root_dom ? mapper.mappedRootIds() : mapper.mappedRFactorIds();
+
+  std::vector<IterDomain*> null_dims;
+  const std::vector<IterDomain*>& projected_dims =
+      projected_dim_map.find(of_tv) == projected_dim_map.end()
+      ? null_dims
+      : projected_dim_map.at(of_tv);
+
+  auto of_tv_root_no_reductions = TensorDomain::noReductions(of_tv_root);
+
+  auto contiguity = of_tv->domain()->contiguity();
+  // Appears after reductions the reduction domain often has a contiguity entry.
+  // This only matters if the result of the reduction is an output
+  if (contiguity.size() == of_tv_root.size() &&
+      contiguity.size() != of_tv_root_no_reductions.size()) {
+    std::vector<bool> new_contiguity;
+    for (auto i : c10::irange(of_tv_root.size())) {
+      if (!of_tv_root[i]->isReduction()) {
+        new_contiguity.push_back(contiguity[i]);
+      }
     }
-    vector_size = next_vector_size;
-    next_vector_size *= 2;
+    contiguity = new_contiguity;
   }
+  of_tv_root = of_tv_root_no_reductions;
 
-  return vector_size;
+  auto of_tv_root_size = of_tv_root.size();
+
+  // Filter out 0-dim tensors
+  if (of_tv_root_size < 1) {
+    return {};
+  }
+
+  // Filter out mismatched contiguity info
+  if (of_tv_root_size != contiguity.size()) {
+    return {};
+  }
+
+  std::vector<Val*> vectorizable_dim_sizes;
+
+  // Order is important, need to make sure dimensions match up correctly with
+  // what was propogated through the mapper. The mapper's dimensions is
+  // propogated in the order of the reference, if that order doesn't match the
+  // tensor we're mapping too then a transpose interfered with expanded the
+  // vectorize dimension.
+  size_t projected_dims_i = projected_dims.size();
+
+  for (auto i : c10::irange(of_tv_root_size)) {
+    if (projected_dims_i == 0) {
+      break;
+    }
+    auto root_i = of_tv_root_size - i - 1;
+    auto root_id = of_tv_root[root_i];
+
+    if (root_id->extent()->isOneInt() || root_id->isBroadcast()) {
+      if (projected_dims[projected_dims_i - 1]->sameAs(root_id)) {
+        --projected_dims_i;
+      }
+      continue;
+    }
+
+    // Not contiguous
+    if (!contiguity[root_i]) {
+      break;
+    }
+
+    // Mapping order isn't correct, cannot expand vectorization dimension.
+    if (!projected_dims[--projected_dims_i]->sameAs(root_id)) {
+      break;
+    }
+
+    vectorizable_dim_sizes.insert(
+        vectorizable_dim_sizes.begin(),
+        mapper.getMaybePartialMappedExtent(root_id));
+    if (mapper.hasPartialExtent(root_id)) {
+      // If we have a partial map we cannot extend the mapping any further for
+      // vectorization.
+      break;
+    }
+  }
+  return vectorizable_dim_sizes;
```

Comment:
Is there any reason why we return a vector of `Val*` instead of returning `SimplifyingIrBuilder::mulExpr` of these vals?

Reply:
No, I just feel strange about adding expressions generally to fusions during heuristics. Maybe I shouldn't.

Reply:
I no longer feel strange about doing that. I've started adding a slew of new expressions 😄